ps sparse rpc

gcramer23 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #58112 DdpSparseDenseRpcTrainer
* #58083 ps batch
* **#58003 ps sparse rpc**

adds trainer class DdpTrainer
adds trainer class DdpSparseRpcTrainer
adds server class ParameterServerBase
adds server class AverageParameterServer
adds experiment ddp_cpu_sparse_rpc_nccl_allreduce
adds experiment ddp_cuda_sparse_rpc_nccl_allreduce

quip document https://fb.quip.com/iQUtAeKIxWpF

Differential Revision: [D29379696](https://our.internmc.facebook.com/intern/diff/D29379696)

gcramer23 committed:
ps sparse rpc

[ghstack-poisoned]

gcramer23 committed:
Update on "ps sparse rpc"




[ghstack-poisoned]

gcramer23 committed:
Update on "ps sparse rpc"




[ghstack-poisoned]

gcramer23 committed:
Update on "ps sparse rpc"




[ghstack-poisoned]

gcramer23 committed:
Update on "ps sparse rpc"




[ghstack-poisoned]

gcramer23 commented on benchmarks/distributed/rpc/parameter_server/servers/AverageParameterServer.py:
```diff
@@ -0,0 +1,98 @@
+    ):
+        super().__init__(rank)
+
+        self.lock = threading.Lock()
```
pull request to fix issue with instance lock https://github.com/pytorch/pytorch/pull/57943

mrshenli replied:
No need to change anything, just FYI. As a temporary solution, IIUC, we can add a `__getstates__` to tell avoid pickling RRefs. 

```python
def __getstates__():
    return {}
```

gcramer23 commented on benchmarks/distributed/rpc/parameter_server/trainers/DdpSparseRpcTrainer.py:
```diff
@@ -0,0 +1,143 @@
+        tensor = bucket.get_tensor()
+        tensors_count = len(cref.bucket_to_parameters(bucket))
+        if tensor.is_sparse:
+            return cref.process_bucket(state, bucket)
```
issue created to track work for fixing RuntimeError https://github.com/pytorch/pytorch/issues/58255

gcramer23 replied:
the issue is closed

gcramer23 reviewed (commented):
I tested before with workarounds with everything passing. I can't test until the pull request for the lock is merged and the issue for RuntimeError is closed.

gcramer23 commented on benchmarks/distributed/rpc/parameter_server/servers/AverageParameterServer.py:
```diff
@@ -0,0 +1,98 @@
+
+        self.rank = rank
+        self.trainer_count = trainer_count
+        self.lc = lc
```
remove line

gcramer23 replied:
done

gcramer23 commented on benchmarks/distributed/rpc/parameter_server/servers/AverageParameterServer.py:
```diff
@@ -0,0 +1,98 @@
+        self,
+        rank,
+        trainer_count,
+        lc,
```
remove arg

gcramer23 replied:
done

gcramer23 committed:
Update on "ps sparse rpc"


adds base trainer class RpcTrainerBase
adds trainer class DdpSparseRpcTrainer
adds base server classes ParameterServerBase and AverageParameterServerBase
adds server class AverageParameterServer
adds experiment ddp_cpu_sparse_rpc_nccl_allreduce
adds experiment ddp_cuda_sparse_rpc_nccl_allreduce

quip document https://fb.quip.com/iQUtAeKIxWpF

[ghstack-poisoned]

mrshenli commented on benchmarks/distributed/rpc/parameter_server/launcher.py:
```diff
@@ -19,9 +19,9 @@
-def get_name(rank, configs):
-    t_count = configs.trainer_count
-    ps_count = configs.ps_count
+def get_name(rank, config):
```
nit: we usually focus on a specific thing in individual PRs. For example, if you would like to modify arg names to simplicity or better readability, those modifications themself can be a separate PR. I am just flagging this. No need to split that from the current PR.

gcramer23 replied:
okay.

mrshenli commented on benchmarks/distributed/rpc/parameter_server/launcher.py:
```diff
@@ -126,6 +127,21 @@ def split_list(arr, n):
+        ps_config = config.ps_config
+        ps = get_benchmark_ps_map()[str(ps_config["ps_class"])]
+        for rank, ps_rref in ps_rrefs.items():
+            metrics = rpc.rpc_sync(
+                ps_rref.owner(),
+                ps.get_metrics_rpc,
+                args=(ps_rref,)
+            )
```
can this be simplified to the following and get rid of the `ps` arg?

```python
ps_rref.rpc_sync().get_metrics_rpc()
```

gcramer23 replied:
yes simplified call and removed `ps` variable.

mrshenli commented on benchmarks/distributed/rpc/parameter_server/launcher.py:
```diff
@@ -105,15 +106,15 @@ def call_trainers(config, model, train_data, parameter_server_rrefs):
         ps = get_benchmark_ps_map()[str(ps_config["ps_class"])]
-    futs = call_trainers(config, model, data, parameter_server_rrefs)
+    futs = call_trainers(config, model, data, ps_rrefs)
     for fut in futs:
         fut.wait()
```
there is an `torch.futures.wait_all()` API for this.

gcramer23 replied:
okay.

mrshenli commented on benchmarks/distributed/rpc/parameter_server/launcher.py:
```diff
@@ -105,15 +106,15 @@ def call_trainers(config, model, train_data, parameter_server_rrefs):
         fut.wait()
-    for ps_rref in parameter_server_rrefs.values():
+    for ps_rref in ps_rrefs.values():
         rpc.rpc_sync(
```
this can be

```python
ps_rref.rpc_sync().reset_state()
```

gcramer23 replied:
okay.

mrshenli commented on benchmarks/distributed/rpc/parameter_server/experiment_scripts/ddp_cpu_sparse_rpc_nccl_allreduce.sh:
```diff
@@ -0,0 +1,11 @@
+cd ..
+
+python -u launcher.py \
+    --benchmark="4" \
```
curious, any reason for using numbers to for `--benchmark` instead of allowing users to directly specify `timeout`, `batch_size`, etc.?

gcramer23 replied:
no reason, the benchmark config can be removed since it all runs use the same arguments.

mrshenli commented on benchmarks/distributed/rpc/parameter_server/configurations/benchmark_configurations.json:
```diff
@@ -4,5 +4,17 @@
         "batch_size": 5
+    },
+    "4": {
+        "trainer_count": 2,
+        "ps_count": 1,
```
nit: DDP launcher names such arguments as `--nnode`, `--nproce_per_node`. We probably wanna stay with the same convention, e.g., `--ntrainer`, `--nps`

https://github.com/pytorch/pytorch/blob/0362b753dba22c1ccab2f9f7ee9f6de2ffe45368/torch/distributed/launch.py#L43

gcramer23 replied:
okay.

mrshenli commented on benchmarks/distributed/rpc/parameter_server/configurations/benchmark_configurations.json:
```diff
@@ -4,5 +4,17 @@
         "ps_count": 0,
         "rpc_async_timeout": 15,
```
is this timeout just for `rpc_async` and not for `rpc_sync` / `remote`? If so, any reason?

gcramer23 replied:
I was using the timeout on the training calls that use rpc_async. I'll move it to the rpc_backend_options.

mrshenli commented on benchmarks/distributed/rpc/parameter_server/configurations/server_configurations.json:
```diff
@@ -0,0 +1,14 @@
+            "use_cuda_rpc": false
+        }
+    },
+    "AverageCudaParameterServer": {
```
curious, why do we need this config file. Is it possible to directly use `AverageCudaParameterServer` class when users pass in `--server=AverageCudaParameterServer`. Is there any reason that we need to use this config as a relay?

gcramer23 replied:
If I remove this config file, then the extensible quality attribute is affected. All parameter servers will have the same arguments given in the initialization of the class. The server class can be passed as a benchmark argument, and the config could be the arguments for the server.

gcramer23 replied:
actually, I can remove the config file but keep the functionality in the launcher. a user needs to specify a config file for additional arguments for their server or trainer.

mrshenli commented on benchmarks/distributed/rpc/parameter_server/launcher.py:
```diff
@@ -185,22 +202,23 @@ def run_benchmark(rank, model, data, config):
+        trainer_configs = config.trainer_config["configurations"]
+        if config.ps_count > 0:
+            ps_configs = config.ps_config["configurations"]
+            if (USE_CUDA_RPC in trainer_configs and
+                trainer_configs[USE_CUDA_RPC] and
+                USE_CUDA_RPC in ps_configs and
+                ps_configs[USE_CUDA_RPC] and
+                    config.ps_count > 0):
```
curious, is this format consistent with PEP8/Flake8? I had an impression that it would generate sth like:

```python
if (
    USE_CUDA_RPC in trainer_configs and
    trainer_configs[USE_CUDA_RPC] and
    USE_CUDA_RPC in ps_configs and
    ps_configs[USE_CUDA_RPC] and
):
```

gcramer23 replied:
yes no warnings/errors after flake8 lint. I'll update the conditional statement.

mrshenli commented on benchmarks/distributed/rpc/parameter_server/launcher.py:
```diff
@@ -185,22 +202,23 @@ def run_benchmark(rank, model, data, config):
+            ps_configs = config.ps_config["configurations"]
+            if (USE_CUDA_RPC in trainer_configs and
+                trainer_configs[USE_CUDA_RPC] and
+                USE_CUDA_RPC in ps_configs and
```
what if only one of trainer and ps wants to use CUDA RPC?

gcramer23 replied:
The benchmark was designed so that all the servers and trainers will either use CUDA RPC or CPU RPC. 

I'll add benchmark arguments for the number of servers and trainers that use CUDA RPC.

mrshenli replied:
I see, this means we should error out when there are conflicts?

gcramer23 replied:
I have added assertions on trainer and server counts.

I haven't been able to test using a mixed configuration. For example 1 cputrainer 1 cpuserver and 1 cudatrainer 1 cudatrainer because of CUDA RPC not implemented error.

mrshenli commented on benchmarks/distributed/rpc/parameter_server/servers/AverageParameterServer.py:
```diff
@@ -0,0 +1,96 @@
+        self = ps_rref.local_value()
+        if type(gradient) is list:
+            gradient = self.sparse_rpc_format_to_tensor(gradient)
+        if not self.use_cuda_rpc:
```
if the purpose of this `if` clause is to avoid unintentional cpu->cuda copy, we can drop the branch and always call `gradient.cuda(self.rank)`, as it will be a no-op if it is already on the target cuda device.

gcramer23 replied:
okay.

mrshenli commented on benchmarks/distributed/rpc/parameter_server/servers/AverageParameterServerBase.py:
```diff
@@ -0,0 +1,14 @@
+from .ParameterServerBase import ParameterServerBase
+
+
+class AverageParameterServerBase(ParameterServerBase):
```
Curious, is there a parameter server that is not averaging gradients

gcramer23 replied:
From my limited reading, it seems that all parameter servers average gradients. I'll remove the AverageParameterServerBase class.

mrshenli commented on benchmarks/distributed/rpc/parameter_server/servers/ParameterServerBase.py:
```diff
@@ -0,0 +1,101 @@
+
+    PARAMETER_SERVER_BATCH_METRIC = "parameter_server_batch_metric"
+    PARAMETER_SERVER_STRAGGLER_METRIC = "parameter_server_straggler_metric"
+    BP_LOC_STRAGGLER = "bp_location_staggler"
```
curious, what does "location" in the name mean?

gcramer23 replied:
I need to update this constant to be `PARAM_LOC_STRAGGLER. 

ddp_communication_hook processes gradient buckets. By flattening and combining all the buckets, you can obtain a single list of parameters. The location for a bucket being processed is the starting parameter number from zero. 

> curious, what does "location" in the name mean?

location + 1 = parameter number of the first parameter in the bucket

The location is used for keys to record metrics.

gcramer23 replied:
I will update `BP_LOC_STRAGGLER` to `PARAM_INDEX_STRAGGLER`

mrshenli commented on benchmarks/distributed/rpc/parameter_server/servers/ParameterServerBase.py:
```diff
@@ -0,0 +1,101 @@
+            key
+        )
+
+    def record_straggler_start(self, key, cuda=True):
```
could you please add some docs to these method to explain what does straggler and batch mean? And what are their start and end time instances?

gcramer23 replied:
okay.

mrshenli commented on benchmarks/distributed/rpc/parameter_server/servers/AverageParameterServer.py:
```diff
@@ -0,0 +1,96 @@
+        self.futures.clear()
+        self.gradient_dict.clear()
+
+    def process_gradient(self, gradient, param_loc):
```
Would I be correct if I assume `param_loc` means `param_index`? Besides is the loc here mean the same thing as the loc in `BP_LOC_STRAGGLER`?

gcramer23 replied:
> Would I be correct if I assume param_loc means param_index?

yes `param_loc` is zero indexed `param_index` if all the params are part of the same list.

> Besides is the loc here mean the same thing as the loc in BP_LOC_STRAGGLER?

yes need to update `BP_LOC_STRAGGLER`

I'll also update loc to index in the relevant variables.

gcramer23 replied:
the parameter index isn't needed. I will use the gradient bucket index.

mrshenli commented on benchmarks/distributed/rpc/parameter_server/trainers/DdpSparseRpcTrainer.py:
```diff
@@ -0,0 +1,149 @@
+    def process_bucket(state, bucket):
+        cref = state.cref
+        tensor = bucket.get_tensor()
+        tensors_count = len(cref.bucket_to_parameters(bucket))
```
do we expect only sparse tensors here or can there be dense tensors as well? If sparse-only, there is always just one tensor in the bucket.

gcramer23 replied:
the process_bucket is designed to process a bucket so it can have sparse or dense tensors.

mrshenli commented on benchmarks/distributed/rpc/parameter_server/trainers/RpcTrainerBase.py:
```diff
@@ -0,0 +1,54 @@
+
+    def send_async_request(self, key, rref, server_method, *args, cuda=True):
+        self.record_rpc_fut_start(key, f"{self.RPC_FUT}_{server_method.__name__}", cuda)
+        fut = rpc.rpc_async(
```
there are RRef helpers to simplify this. I.e.:

```python
rref.rpc_async().server_method(*args)
```

gcramer23 replied:
okay will update. I might end up moving this method.

gcramer23 replied:
removed the method.

mrshenli commented on benchmarks/distributed/rpc/parameter_server/trainers/RpcTrainerBase.py:
```diff
@@ -0,0 +1,54 @@
+    def __init__(self, rank):
+        super().__init__(rank)
+
+    def send_async_request(self, key, rref, server_method, *args, cuda=True):
```
the name does not seem to indicate what this is doing. IIUC, the main purpose of this wrapper is to profile the delay?

gcramer23 replied:
it sends the request and adds a metric for the future delay. it is a bit verbose for sending an async request. I will get rid of the method.

mrshenli commented on benchmarks/distributed/rpc/parameter_server/trainers/RpcTrainerBase.py:
```diff
@@ -0,0 +1,54 @@
+
+        return fut.then(request_time_callback)
+
+    def send_sync_request(self, key, rref, server_method, *args, cuda=True):
```
this API is not used in this PR. Let's only introduce new API when they are needed.

gcramer23 replied:
okay will remove.

gcramer23 commented on benchmarks/distributed/rpc/parameter_server/servers/ParameterServerBase.py:
```diff
@@ -0,0 +1,101 @@
+    def clear_metrics(self):
+        return self.__metrics_logger.clear_metrics()
+
+    def sparse_tensor_to_rpc_format(self, sparse_tensor):
```
I will move it to a utils file since the RPC trainers also use identical methods.

Should I update the trainers in a different pr? I also want to remove duplicate code from the DdpNcclTrainer since the training loop is identical to DdpSpraseRpcTrainer because gradients are not being updated locally after .backward()

gcramer23 commented on benchmarks/distributed/rpc/parameter_server/trainers/DdpSparseRpcTrainer.py:
```diff
@@ -0,0 +1,149 @@
+        self.rank = rank
+        self.trainer_count = trainer_count
+        self.ps_rref = ps_rref
+        self.ps = self.PS_MAP[ps_name]
```
How the trainer is currently set up, I need it to obtain the average gradient method. However, I will change the trainer to call using RPC helper methods, so I don't need it anymore.

gcramer23 commented on benchmarks/distributed/rpc/parameter_server/trainers/DdpSparseRpcTrainer.py:
```diff
@@ -0,0 +1,149 @@
+    class HookState:
+
+        def __init__(self, cref, process_group):
+            self.cref = cref
```
it's a reference to the self keyword of the trainer instance.

gcramer23 commented on benchmarks/distributed/rpc/parameter_server/trainers/DdpSparseRpcTrainer.py:
```diff
@@ -0,0 +1,149 @@
+
+    class HookState:
+
+        def __init__(self, cref, process_group):
```
okay.

gcramer23 commented on benchmarks/distributed/rpc/parameter_server/trainers/DdpSparseRpcTrainer.py:
```diff
@@ -0,0 +1,149 @@
+        process_group_size = self.trainer_count
+
+        store = c10d.FileStore("/tmp/tmpn_k_8so02", process_group_size)
+        if self.backend == c10d.Backend.GLOO:
+            process_group = c10d.ProcessGroupGloo(store, self.rank, process_group_size)
+        elif self.backend == c10d.Backend.NCCL:
+            process_group = c10d.ProcessGroupNCCL(store, self.rank, process_group_size)
```
okay ill change it.

gcramer23 commented on benchmarks/distributed/rpc/parameter_server/trainers/DdpSparseRpcTrainer.py:
```diff
@@ -0,0 +1,149 @@
+        for i in range(len(data)):
+            data[i][0] = data[i][0].cuda(self.rank)
+            data[i][1] = data[i][1].cuda(self.rank)
+        torch.cuda.synchronize(self.rank)
```
okay.  I don't know if I need this. I might remove it after checking to see how tensors are moved to the device.

gcramer23 replied:
removed this since tensor from cpu to gpu is blocking.

gcramer23 commented on benchmarks/distributed/rpc/parameter_server/trainers/DdpSparseRpcTrainer.py:
```diff
@@ -0,0 +1,149 @@
+        "AverageParameterServer": AverageParameterServer
+    }
+
+    class HookState:
```
yes that is correct. Gradients are not updated locally after the .backward(), so the HookState class is identical to DdpNcclTrainer HookState. I will move the HookState class to a base trainer class.

gcramer23 commented on benchmarks/distributed/rpc/parameter_server/trainers/DdpSparseRpcTrainer.py:
```diff
@@ -0,0 +1,149 @@
+        self.epochs = epochs
+
+    @staticmethod
+    def get_tensor_fut(bucket):
```
do not need it. that is dead code.

gcramer23 commented on benchmarks/distributed/rpc/parameter_server/trainers/DdpSparseRpcTrainer.py:
```diff
@@ -0,0 +1,149 @@
+        cref = state.cref
+        tensor = bucket.get_tensor()
+        tensors_count = len(cref.bucket_to_parameters(bucket))
+        sparse = tensor.is_sparse
```
from old logic didn't think to remove the variable. I will remove it.

gcramer23 commented on benchmarks/distributed/rpc/parameter_server/test_scripts.py:
```diff
@@ -21,3 +21,11 @@ def run_script(script_name):
 
 def test_ddp_nccl_allreduce():
```
> are these tests linked to CI?

I did not link them to the CI. Are they supposed to be added to CI? it is a benchmark, and I do not see other benchmarks with test files.

> how do we test them today?

I am running them locally on a node when I want to verify experiments are working.

mrshenli replied:
Some other benchmarks have counterpart tests in CI. We can leave this as-is for now, and figure out how we should deal with it later.

gcramer23 replied:
okay

H-Huang commented on benchmarks/distributed/rpc/parameter_server/trainers/DdpSparseRpcTrainer.py:
```diff
@@ -0,0 +1,149 @@
+        return DdpSparseRpcTrainer.hook
+
+    def train(self, model, data):
+        torch.manual_seed(0)
```
If we are setting a seed, then maybe this should be done in launcher? Also don't we also need to seed cuda?

gcramer23 replied:
add seed to launcher. also, added backend reproducibility.
```python
    torch.manual_seed(args.torch_seed)
    torch.cuda.manual_seed_all(args.cuda_seed)
    torch.backends.cudnn.benchmark = True
    torch.backends.cudnn.deterministic = True
```

H-Huang commented on benchmarks/distributed/rpc/parameter_server/launcher.py:
```diff
@@ -30,30 +30,31 @@ def get_name(rank, configs):
+    rem = config.trainer_count % config.ps_count
+    if ps_rank - config.trainer_count < rem:
         ps_trainer_count += 1
+    ps_trainer_count = int(ps_trainer_count)
```
nit: you can just make line 48 `ps_trainer_count = config.trainer_count // config.ps_count` to accomplish the same thing

Question, do we have any restrictions on the # of parameter server and trainer counts?

gcramer23 replied:
no restriction on server and trainer counts. this benchmark is set to run within a single node. the dev queue has 4 gpus per node. train queue has 8 gpus per node. 

I have the benchmark set so that the number of trainers must be divisible by the number of servers.

gcramer23 commented on benchmarks/distributed/rpc/parameter_server/configurations/benchmark_configurations.json:
```diff
@@ -4,5 +4,17 @@
+        "rpc_async_timeout": 30,
+        "batch_size": 10
+    },
+    "5": {
```
I removed this file in the updates.

gcramer23 replied:
I removed this file. all these arguments in the configurations have been moved to the launcher arguments. I added a help section to each argument in case the name is not self explanatory.

gcramer23 committed:
Update on "ps sparse rpc"


adds base trainer class RpcTrainerBase
adds trainer class DdpSparseRpcTrainer
adds base server classes ParameterServerBase and AverageParameterServerBase
adds server class AverageParameterServer
adds experiment ddp_cpu_sparse_rpc_nccl_allreduce
adds experiment ddp_cuda_sparse_rpc_nccl_allreduce

quip document https://fb.quip.com/iQUtAeKIxWpF

[ghstack-poisoned]

gcramer23 commented:
issue for sparse tensor storage access when using CUDA RPC https://github.com/pytorch/pytorch/issues/58755

gcramer23 committed:
Update on "ps sparse rpc"


adds base trainer class RpcTrainerBase
adds trainer class DdpSparseRpcTrainer
adds base server classes ParameterServerBase and AverageParameterServerBase
adds server class AverageParameterServer
adds experiment ddp_cpu_sparse_rpc_nccl_allreduce
adds experiment ddp_cuda_sparse_rpc_nccl_allreduce

quip document https://fb.quip.com/iQUtAeKIxWpF

[ghstack-poisoned]

gcramer23 committed:
Update on "ps sparse rpc"


adds trainer class DdpTrainer
adds trainer class DdpRpcTrainer
adds trainer class DdpSparseRpcTrainer
adds base server classes ParameterServerBase
adds server class AverageParameterServer
adds experiment ddp_cpu_sparse_rpc_nccl_allreduce
adds experiment ddp_cuda_sparse_rpc_nccl_allreduce

quip document https://fb.quip.com/iQUtAeKIxWpF

[ghstack-poisoned]

gcramer23 committed:
Update on "ps sparse rpc"


adds trainer class DdpTrainer
adds trainer class DdpRpcTrainer
adds trainer class DdpSparseRpcTrainer
adds base server classes ParameterServerBase
adds server class AverageParameterServer
adds experiment ddp_cpu_sparse_rpc_nccl_allreduce
adds experiment ddp_cuda_sparse_rpc_nccl_allreduce

quip document https://fb.quip.com/iQUtAeKIxWpF

[ghstack-poisoned]

mrshenli commented on benchmarks/distributed/rpc/parameter_server/servers/AverageParameterServer.py:
```diff
@@ -0,0 +1,142 @@
+        r"""
+        a parameter server that averages the gradients
+        from trainers for each training iteration step.
+        gradients are added as they are received from trainers.
```
nit: capital letter in the beginning of a sentence.

gcramer23 replied:
okay.

mrshenli commented on benchmarks/distributed/rpc/parameter_server/trainers/DdpTrainer.py:
```diff
@@ -0,0 +1,167 @@
+        r"""
+        performs an iteration of training.
+        Args:
+            ddp_model (object): distributed data parallel model
```
`ddp_model` should be an `nn.Module` instance?

gcramer23 replied:
okay.

mrshenli commented on benchmarks/distributed/rpc/parameter_server/trainers/DdpTrainer.py:
```diff
@@ -0,0 +1,167 @@
+        Args:
+            ddp_model (object): distributed data parallel model
+            criterion (object): loss function to measure model
+            optimizer (object): updates model parameters
```
`optimizer` should be an `optim.Optimizer` instance?

gcramer23 replied:
okay.

mrshenli commented on benchmarks/distributed/rpc/parameter_server/trainers/DdpSparseRpcTrainer.py:
```diff
@@ -0,0 +1,53 @@
+            trainer_count (int): count of trainer in the world
+            process_group (object): distributed process group
+            use_cuda_rpc (bool): indicator for CUDA RPC
+            server_rref (object): remote reference to the server
```
This should be an RRef instance?

gcramer23 replied:
okay.

mrshenli commented on benchmarks/distributed/rpc/parameter_server/trainers/DdpSparseRpcTrainer.py:
```diff
@@ -0,0 +1,53 @@
+        cref = state.cref
+        tensor = bucket.get_tensor()
+        if tensor.is_sparse:
+            return cref.process_bucket(state, bucket)
```
Does `process_bucket` always sync buckets using RPC? If yes, let make it clear in the name.

gcramer23 replied:
yes it always uses RPC. I will update the name to process_bucket_with_server.

mrshenli commented on benchmarks/distributed/rpc/parameter_server/trainers/DdpRpcHelper.py:
```diff
@@ -0,0 +1,48 @@
+from utils import sparse_tensor_to_rpc_format, sparse_rpc_format_to_tensor
+
+
+class DdpRpcHelper:
```
curious, why does this needs to be a class instead of a helper function in utils?

gcramer23 replied:
I think that the process_bucket_with_remote_server is specific to the class hierarchy of the current trainer implementations. It can be a helper method in utils. I will update the code.

gcramer23 committed:
Update on "ps sparse rpc"


adds trainer class DdpTrainer
adds trainer class DdpRpcTrainer
adds trainer class DdpSparseRpcTrainer
adds base server classes ParameterServerBase
adds server class AverageParameterServer
adds experiment ddp_cpu_sparse_rpc_nccl_allreduce
adds experiment ddp_cuda_sparse_rpc_nccl_allreduce

quip document https://fb.quip.com/iQUtAeKIxWpF

[ghstack-poisoned]

gcramer23 committed:
Update on "ps sparse rpc"


adds trainer class DdpTrainer
adds trainer class DdpRpcTrainer
adds trainer class DdpSparseRpcTrainer
adds base server classes ParameterServerBase
adds server class AverageParameterServer
adds experiment ddp_cpu_sparse_rpc_nccl_allreduce
adds experiment ddp_cuda_sparse_rpc_nccl_allreduce

quip document https://fb.quip.com/iQUtAeKIxWpF

[ghstack-poisoned]

gcramer23 commented:
@gcramer23 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29379696).
